// ng2-file-input/dist/bundles/ng2-file-input.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@angular/common'), require('bergben-angular2-file-drop')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/platform-browser', '@angular/common', 'bergben-angular2-file-drop'], factory) :
	(factory((global['ng2-file-input'] = global['ng2-file-input'] || {}),global.ng.core,global._angular_platformBrowser,global._angular_common,global.bergbenAngular2FileDrop));
}(this, (function (exports,_angular_core,_angular_platformBrowser,_angular_common,bergbenAngular2FileDrop) { 'use strict';

var Ng2FileInputOptions = (function () {
    function Ng2FileInputOptions() {
        this.dropText = "Drop file here";
        this.browseText = "Browse";
        this.removeText = "Remove";
        this.invalidFileText = "You have picked an invalid or disallowed file.";
        this.invalidFileTimeout = 8000;
        this.removable = true;
        this.multiple = false;
        this.showPreviews = true;
        this.extensions = [];
    }
    return Ng2FileInputOptions;
}());

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (undefined && undefined.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
exports.Ng2FileInputComponent = (function () {
    function Ng2FileInputComponent(sanitizer, defaultOptions) {
        this.sanitizer = sanitizer;
        this.defaultOptions = defaultOptions;
        this.alreadyEmitted = false;
        this.fileIsOver = false;
        this.invalidFile = false;
        this.multiple = null;
        this.removable = null;
        this.showPreviews = null;
        this.output = new _angular_core.EventEmitter();
        this.currentFiles = [];
    }
    Ng2FileInputComponent.prototype.ngOnInit = function () {
        this.dropText = this.dropText || this.defaultOptions.dropText;
        this.browseText = this.browseText || this.defaultOptions.browseText;
        this.removeText = this.removeText || this.defaultOptions.removeText;
        this.invalidFileText = this.invalidFileText || this.defaultOptions.invalidFileText;
        this.invalidFileTimeout = this.invalidFileTimeout || this.defaultOptions.invalidFileTimeout;
        this.multiple = this.multiple !== null ? this.multiple : this.defaultOptions.multiple;
        this.removable = this.removable !== null ? this.removable : this.defaultOptions.removable;
        this.showPreviews = this.showPreviews !== null ? this.showPreviews : this.defaultOptions.showPreviews;
        this.extensions = this.extensions || this.defaultOptions.extensions;
    };
    Ng2FileInputComponent.prototype.fileOver = function (fileIsOver) {
        this.fileIsOver = fileIsOver;
    };
    Ng2FileInputComponent.prototype.onFileDrop = function (file) {
        var _this = this;
        if (!this.multiple) {
            //make sure only to emit once
            if (!this.alreadyEmitted) {
                this.alreadyEmitted = true;
                this.handleFile(file);
                setTimeout(function () { _this.alreadyEmitted = false; }, 0);
            }
        }
        else {
            this.handleFile(file);
        }
    };
    Ng2FileInputComponent.prototype.fileSelected = function (event) {
        this.invalidFile = false;
        var files = event.target.files;
        if (files && files.length) {
            if (!this.multiple) {
                //make sure only to emit one
                this.handleFile(files[0]);
            }
            else {
                for (var i = 0; i < files.length; i++) {
                    this.handleFile(files[i]);
                }
            }
        }
        setTimeout(function () { event.target.value = ""; }, 0);
    };
    Ng2FileInputComponent.prototype.removeFile = function (file) {
        if (this.removable) {
            for (var i = 0; i < this.currentFiles.length; i++) {
                if (this.currentFiles[i] === file) {
                    this.currentFiles.splice(i, 1);
                    this.output.emit({
                        currentFiles: this.currentFiles,
                        action: "removed",
                        file: file
                    });
                }
            }
        }
    };
    Ng2FileInputComponent.prototype.getObjectUrl = function (file) {
        return this.sanitizer.bypassSecurityTrustResourceUrl(window.URL.createObjectURL(file));
    };
    Ng2FileInputComponent.prototype.handleFile = function (file) {
        if (this.isValidFile(file)) {
            if (!this.multiple) {
                this.currentFiles = [];
            }
            this.currentFiles.push(file);
            this.output.emit({
                currentFiles: this.currentFiles,
                action: "added",
                file: file
            });
        }
    };
    Ng2FileInputComponent.prototype.isValidFile = function (file) {
        var _this = this;
        if (this.extensions.length > 0) {
            var ext = file.name.split('.').pop();
            if ((this.extensions.indexOf('image/jpg') !== -1) && (this.extensions.indexOf('image/jpeg') === -1)) {
                this.extensions.push('image/jpeg');
            }
            
            if (this.extensions.indexOf(file.type) === -1 && this.extensions.indexOf(ext) === -1) {
                this.invalidFile = true;
                if (this.invalidFileTimeout !== 0) {
                    setTimeout(function () {
                        _this.invalidFile = false;
                    }, this.invalidFileTimeout);
                }
                return false;
            }
        }
        return true;
    };
    return Ng2FileInputComponent;
}());
__decorate([
    _angular_core.Input('drop-text')
], exports.Ng2FileInputComponent.prototype, "dropText", void 0);
__decorate([
    _angular_core.Input('browse-text')
], exports.Ng2FileInputComponent.prototype, "browseText", void 0);
__decorate([
    _angular_core.Input('remove-text')
], exports.Ng2FileInputComponent.prototype, "removeText", void 0);
__decorate([
    _angular_core.Input('invalid-file-text')
], exports.Ng2FileInputComponent.prototype, "invalidFileText", void 0);
__decorate([
    _angular_core.Input('invalid-file-timeout')
], exports.Ng2FileInputComponent.prototype, "invalidFileTimeout", void 0);
__decorate([
    _angular_core.Input('multiple')
], exports.Ng2FileInputComponent.prototype, "multiple", void 0);
__decorate([
    _angular_core.Input('removable')
], exports.Ng2FileInputComponent.prototype, "removable", void 0);
__decorate([
    _angular_core.Input('show-previews')
], exports.Ng2FileInputComponent.prototype, "showPreviews", void 0);
__decorate([
    _angular_core.Input('extensions')
], exports.Ng2FileInputComponent.prototype, "extensions", void 0);
__decorate([
    _angular_core.Output('onChange')
], exports.Ng2FileInputComponent.prototype, "output", void 0);
exports.Ng2FileInputComponent = __decorate([
    _angular_core.Component({
        selector: 'ng2-file-input',
        template: "<div class=\"ng2-file-input\">\n                    <div class=\"ng2-file-input-invalid text-danger\" [hidden]=\"!invalidFile\" [innerHTML]=\"invalidFileText\"></div>\n                    <div fileDrop class=\"ng2-file-input-drop-container\" [ngClass]=\"{'file-is-over': fileIsOver}\" (fileOver)=\"fileOver($event)\"\n                        (onFileDrop)=\"onFileDrop($event)\">\n                        <span [innerHTML]=\"dropText\"></span>\n                        <button type=\"button\" (click)=\"ng2FileInputSelect.click()\" class=\"btn btn-primary\" [innerHTML]=\"browseText\"></button>\n                    </div>\n                    <div class=\"ng2-file-input-files\" *ngIf=\"showPreviews\">\n                        <div *ngFor=\"let file of currentFiles\" class=\"ng2-file-input-file\" [ngClass]=\"{'image':file.type.indexOf('image')!==-1}\">\n                            <span [innerHTML]=\"file.name\" class=\"ng2-file-input-file-text\"></span>\n                            <img [src]=\"getObjectUrl(file)\" *ngIf=\"file.type.indexOf('image')!==-1\">\n                            <span class=\"ng2-file-input-file-text remove\" (click)=\"removeFile(file)\" *ngIf=\"removable\">\n                                <p [innerHTML]=\"removeText\"></p>\n                            </span> \n                        </div>\n                    </div>\n                    <input type=\"file\" #ng2FileInputSelect (change)=\"fileSelected($event)\" [attr.multiple]=\"multiple ? true : null\">\n                </div>",
    }),
    __param(0, _angular_core.Inject(_angular_core.forwardRef(function () { return _angular_platformBrowser.DomSanitizer; }))), __param(1, _angular_core.Inject(_angular_core.forwardRef(function () { return Ng2FileInputOptions; })))
], exports.Ng2FileInputComponent);

var __decorate$1 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var USER_OPTIONS = new _angular_core.OpaqueToken('ng2 file input custom user options');
function optionsFactory(userOptions) {
    var options = new Ng2FileInputOptions();
    Object.assign(options, userOptions);
    return options;
}
exports.Ng2FileInputModule = Ng2FileInputModule_1 = (function () {
    function Ng2FileInputModule() {
    }
    Ng2FileInputModule.forRoot = function (options) {
        if (options === void 0) { options = {}; }
        return {
            ngModule: Ng2FileInputModule_1,
            providers: [{
                    provide: USER_OPTIONS,
                    useValue: options
                }, {
                    provide: Ng2FileInputOptions,
                    useFactory: optionsFactory,
                    deps: [USER_OPTIONS]
                }]
        };
    };
    return Ng2FileInputModule;
}());
exports.Ng2FileInputModule = Ng2FileInputModule_1 = __decorate$1([
    _angular_core.NgModule({
        imports: [
            _angular_common.CommonModule,
            bergbenAngular2FileDrop.FileDropModule
        ],
        declarations: [exports.Ng2FileInputComponent],
        exports: [exports.Ng2FileInputComponent]
    })
], exports.Ng2FileInputModule);
var Ng2FileInputModule_1;

exports.Ng2FileInputOptions = Ng2FileInputOptions;

Object.defineProperty(exports, '__esModule', { value: true });

})));
