// ng2-materialize/inline-template/card/card.component.ts
import {
  AfterViewInit,
  Component,
  Directive,
  ElementRef,
  Input,
  ViewChild } from '@angular/core';

@Component({
  selector: 'mz-card',
  template: `<div class="card {{ backgroundClass }}" [class.hoverable]="hoverable"><div class="card-content {{ textClass }}"><span class="card-title"><ng-content select="mz-card-title"></ng-content></span><p><ng-content select="mz-card-content"></ng-content></p></div><div #cardActionWrapper class="card-action"><ng-content select="mz-card-action"></ng-content></div></div>`,
  styles: [``],
})
export class MzCardComponent implements AfterViewInit {
  @Input() backgroundClass: string;
  @Input() hoverable: boolean;
  @Input() textClass: string;

  @ViewChild('cardActionWrapper') cardActionWrapper: ElementRef;

  constructor() { }

  ngAfterViewInit() {
    this.initCardAction();
  }

  private initCardAction() {
    const cardActionElement = $(this.cardActionWrapper.nativeElement).find('mz-card-action');

    // hide action wrapper when mz-card-action element is not present
    // or when mz-card-action does not contain any html (text or elements)
    if (cardActionElement.length === 0 || cardActionElement[0].innerHTML.trim() === '') {
      this.cardActionWrapper.nativeElement.style.display = 'none';
    }
  }
}


// Declare the tags to avoid error: '<mz-card-x>' is not a known element
// https://github.com/angular/angular/issues/11251
// tslint:disable: directive-selector
@Directive({ selector: 'mz-card-title' }) export class MzCardTitleDirective { }
@Directive({ selector: 'mz-card-content' }) export class MzCardContentDirective { }
@Directive({ selector: 'mz-card-action' }) export class MzCardActionDirective { }
